api/v1/storagepool.go

```diff
@@ -0,0 +1,268 @@
+package v1
+
+import (
+	"fmt"
+	"path"
+	"reflect"
+	"regexp"
+	"strconv"
+	"strings"
+
+	lclient "github.com/LINBIT/golinstor/client"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+)
+
+type LinstorStoragePool struct {
+	// Name of the storage pool in linstor.
+	//+kubebuilder:validation:MinLength=3
+	Name string `json:"name"`
+
+	// Properties to set on the storage pool.
+	// +listType=map
+	// +listMapKey=name
+	// +patchMergeKey=name
+	// +patchStrategy=merge
+	Properties []LinstorNodeProperty `json:"properties,omitempty"`
+
+	Lvm     *LinstorStoragePoolLvm     `json:"lvm,omitempty"`
+	LvmThin *LinstorStoragePoolLvmThin `json:"lvmThin,omitempty"`
+
+	Source *LinstorStoragePoolSource `json:"source,omitempty"`
+}
+
+func (p *LinstorStoragePool) ProviderKind() lclient.ProviderKind {
+	switch {
+	case p.Lvm != nil:
+		return lclient.LVM
+	case p.LvmThin != nil:
+		return lclient.LVM_THIN
+	}
+
+	return ""
+}
+
+func (p *LinstorStoragePool) PoolName() string {
+	switch {
+	case p.Lvm != nil:
+		if p.Lvm.VolumeGroup != "" {
+			return p.Lvm.VolumeGroup
+		}
+
+		return p.Name
+	case p.LvmThin != nil:
+		lvName := p.LvmThin.ThinPool
+		if lvName == "" {
+			lvName = p.Name
+		}
+
+		vgName := p.LvmThin.VolumeGroup
+		if vgName == "" {
+			vgName = fmt.Sprintf("linstor_%s", lvName)
+		}
+
+		return fmt.Sprintf("%s/%s", vgName, lvName)
+	}
+	return ""
+}
+
+type LinstorStoragePoolLvm struct {
+	VolumeGroup string `json:"volumeGroup,omitempty"`
+}
+
+type LinstorStoragePoolLvmThin struct {
+	VolumeGroup string `json:"volumeGroup,omitempty"`
+	// ThinPool is the name of the thinpool LV (without VG prefix).
+	ThinPool string `json:"thinPool,omitempty"`
+}
+
+type LinstorStoragePoolSource struct {
+	// HostDevices is a list of device paths used to configure the given pool.
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:validation:MinItems:=1
+	HostDevices []string `json:"hostDevices,omitempty"`
+}
+
+var (
+	SPRegexp = regexp.MustCompile("^[A-Za-z0-9][A-Za-z0-9_-]{1,46}[A-Za-z0-9]$")
+	VGRegexp = regexp.MustCompile("^[A-Za-z0-9.+_-]+$")
+)
+
+func ValidateStoragePools(curSPs, oldSPs []LinstorStoragePool, fieldPrefix *field.Path) field.ErrorList {
+	var result field.ErrorList
+
+	spNames := sets.NewString()
+	devNames := sets.NewString()
+
+	for i := range curSPs {
+		curSP := &curSPs[i]
+		if !SPRegexp.MatchString(curSP.Name) {
+			result = append(result, field.Invalid(
+				fieldPrefix.Child(strconv.Itoa(i), "name"),
+				curSP.Name,
+				"Not a valid LINSTOR Storage Pool name",
+			))
+		}
+
+		if spNames.Has(curSP.Name) {
+			result = append(result, field.Duplicate(
+				fieldPrefix.Child(strconv.Itoa(i), "name"),
+				curSP.Name,
+			))
+		}
+
+		spNames.Insert(curSP.Name)
+		var oldSP *LinstorStoragePool
+		for j := range oldSPs {
+			if oldSPs[j].Name == curSP.Name {
+				oldSP = &oldSPs[j]
+				break
+			}
+		}
+
+		switch {
```

Comment:
This validates that at least one of `curSP.LvmThin` and `curSP.Lvm` are set, but not that at most one is.

Reply:
Good catch. Added a test to cover that scenario